File: server/parcle.js
/**
 * parcle.js — Parcle Memory API helper
 * Base URL: https://api.parcle.ai
 * Docs:     https://docs.parcle.ai/api/memory-api.html
 *
 * All three functions handle errors gracefully:
 *   - saveMemory   → returns the Parcle session_id, or null on failure
 *   - queryMemory  → returns top-5 result objects, or [] on failure
 *   - listRecentMemories → returns sources sorted newest-first, or [] on failure
 */

require('dotenv').config({ path: '../.env' });
const axios = require('axios');

const BASE_URL = 'https://api.parcle.ai';

// The user_id scopes all memory under a single namespace for this project.
// You can make this dynamic per-user if needed later.
const DEFAULT_USER_ID = process.env.ENTER_PROJECT_ID || 'eternal-architect';

/**
 * Returns an Axios instance pre-configured with the Parcle auth header.
 * Throws clearly if the API key is missing.
 */
function createClient() {
  const apiKey = process.env.PARCLE_API_KEY;
  if (!apiKey) {
    throw new Error('PARCLE_API_KEY is not set in environment variables.');
  }
  return axios.create({
    baseURL: BASE_URL,
    headers: {
      Authorization: `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
    },
    timeout: 30_000,
  });
}

/**
 * Ensures the user namespace exists in Parcle before writing to it.
 * This is idempotent — calling it on an existing user is safe.
 */
async function ensureUser(client, userId) {
  try {
    await client.post('/v1/users', {
      user_id: userId,
      name: 'Eternal Architect Agent',
      timezone: 'UTC',
    });
  } catch (err) {
    // 409 / already exists is fine; surface real errors
    if (err?.response?.status !== 409) {
      console.warn('[Parcle] ensureUser warning:', err?.response?.data || err.message);
    }
  }
}

// ─────────────────────────────────────────────────────────────
// 1. saveMemory(title, content, tags[])
//    Stores a decision, bugfix, or feature note in Parcle memory.
//    Uses ingest_dialog so it's immediately searchable.
// ─────────────────────────────────────────────────────────────
async function saveMemory(title, content, tags = []) {
  const timestamp = new Date().toISOString();
  const projectName = process.env.ENTER_PROJECT_ID || 'eternal-architect';

  console.log(`[Parcle] 💾 saveMemory → title="${title}" tags=${JSON.stringify(tags)}`);

  try {
    const client = createClient();
    const userId = DEFAULT_USER_ID;

    await ensureUser(client, userId);

    // Build a tag object from the tags array + required fields
    const tagObject = {
      project: projectName,
      timestamp,
    };
    tags.forEach((t) => {
      // Accept "key:value" pairs or plain strings stored as labels
      if (typeof t === 'string' && t.includes(':')) {
        const [k, ...rest] = t.split(':');
        tagObject[k.trim()] = rest.join(':').trim();
      } else if (typeof t === 'string') {
        tagObject[t] = 'true';
      } else if (typeof t === 'object') {
        Object.assign(tagObject, t);
      }
    });

    // Encode the memory as a two-turn dialog:
    //   user turn  → the title / question
    //   assistant  → the full content / answer
    const response = await client.post('/v1/memories/ingest_dialog', {
      user_id: userId,
      messages: [
        { role: 'user', content: `[Memory] ${title}` },
        { role: 'assistant', content },
      ],
      tag: tagObject,
    });

    const { session_id, event_id } = response.data;
    console.log(`[Parcle] ✅ Memory saved → session_id=${session_id} event_id=${event_id}`);
    return session_id;
  } catch (err) {
    console.error('[Parcle] ❌ saveMemory failed:', err?.response?.data || err.message);
    return null;
  }
}

// ─────────────────────────────────────────────────────────────
// 2. queryMemory(naturalLanguageQuery)
//    Searches Parcle with a plain-English question.
//    Returns an array of up to 5 result objects shaped as:
//      { answer, confidence, citations, query }
//    The search endpoint returns an SSE stream; we parse it here.
// ─────────────────────────────────────────────────────────────
async function queryMemory(naturalLanguageQuery) {
  console.log(`[Parcle] 🔍 queryMemory → "${naturalLanguageQuery}"`);

  try {
    const apiKey = process.env.PARCLE_API_KEY;
    if (!apiKey) throw new Error('PARCLE_API_KEY is not set.');

    // The search endpoint streams Server-Sent Events.
    // We request it as a stream and parse it chunk-by-chunk, destroying the stream as soon as we get the final data.
    const response = await axios.post(
      `${BASE_URL}/v1/memories/search`,
      {
        user_id: DEFAULT_USER_ID,
        query: naturalLanguageQuery,
      },
      {
        headers: {
          Authorization: `Bearer ${apiKey}`,
          'Content-Type': 'application/json',
          Accept: 'text/event-stream',
        },
        responseType: 'stream',
        timeout: 30_000,
      }
    );

    const stream = response.data;
    let buffer = '';
    let finalData = null;

    const resultPromise = new Promise((resolve, reject) => {
      let seenFinalEvent = false;

      stream.on('data', (chunk) => {
        buffer += chunk.toString('utf8');
        const lines = buffer.split('\n');
        
        // Keep the last partial line in the buffer
        buffer = lines.pop() || '';

        for (const line of lines) {
          const trimmed = line.trim();
          if (trimmed === 'event: final') {
            seenFinalEvent = true;
            continue;
          }
          if (seenFinalEvent && trimmed.startsWith('data:')) {
            try {
              finalData = JSON.parse(trimmed.slice(5).trim());
            } catch (err) {
              console.warn('[Parcle] Could not parse final SSE data:', trimmed);
            }
            // Got the final data, destroy stream and resolve
            stream.destroy();
            resolve(finalData);
            return;
          }
        }
      });

      stream.on('end', () => {
        resolve(finalData);
      });

      stream.on('error', (err) => {
        reject(err);
      });
    });

    finalData = await resultPromise;

    if (!finalData) {
      console.warn('[Parcle] No final event found in search SSE response.');
      return [];
    }

    console.log(
      `[Parcle] ✅ queryMemory result → confidence=${finalData.confidence} citations=${finalData.citations?.length ?? 0}`
    );

    // Return as an array of up to 5 enriched result objects
    const results = [
      {
        query: naturalLanguageQuery,
        answer: finalData.answer,
        confidence: finalData.confidence,
        citations: finalData.citations ?? [],
      },
    ];

    // If citations reference multiple sources, expand them into separate entries
    if (finalData.citations && finalData.citations.length > 1) {
      const extras = finalData.citations.slice(1, 5).map((citation) => ({
        query: naturalLanguageQuery,
        answer: finalData.answer,
        confidence: finalData.confidence,
        citations: [citation],
      }));
      results.push(...extras);
    }

    return results.slice(0, 5);
  } catch (err) {
    console.error('[Parcle] ❌ queryMemory failed:', err?.response?.data || err.message);
    return [];
  }
}

// ─────────────────────────────────────────────────────────────
// 3. listRecentMemories(limit = 10)
//    Fetches the most recent ingested sessions from Parcle,
//    sorted newest-first (order: "desc").
// ─────────────────────────────────────────────────────────────
async function listRecentMemories(limit = 10) {
  console.log(`[Parcle] 📋 listRecentMemories → limit=${limit}`);

  try {
    const client = createClient();

    const response = await client.post('/v1/memories/sources', {
      user_id: DEFAULT_USER_ID,
      type: 'session',
      order: 'desc',
      limit,
      page: 1,
    });

    const { sources = [], total } = response.data;

    console.log(`[Parcle] ✅ listRecentMemories → ${sources.length}/${total} sources returned`);

    // Sort by updated_at descending (API already does this, but be explicit)
    const sorted = [...sources].sort(
      (a, b) => new Date(b.updated_at) - new Date(a.updated_at)
    );

    return sorted;
  } catch (err) {
    console.error('[Parcle] ❌ listRecentMemories failed:', err?.response?.data || err.message);
    return [];
  }
}

module.exports = { saveMemory, queryMemory, listRecentMemories };

File: backend/parcle.py
import os
import json
import asyncio
import urllib.request
import urllib.error
from datetime import datetime

BASE_URL = 'https://api.parcle.ai'
DEFAULT_USER_ID = os.getenv("ENTER_PROJECT_ID", "eternal-architect")

def get_api_key():
    from dotenv import dotenv_values
    env_path = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), '.env')
    config = dotenv_values(env_path)
    api_key = config.get("PARCLE_API_KEY") or os.getenv("PARCLE_API_KEY")
    if not api_key:
        raise ValueError("PARCLE_API_KEY is not set in environment variables.")
    return api_key.strip()

def _make_request(url, payload, method="POST"):
    """Synchronous urllib helper — call via run_in_executor for async use."""
    api_key = get_api_key()
    data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=data,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
            "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/111.0"
        },
        method=method
    )
    with urllib.request.urlopen(req, timeout=10) as resp:
        return json.loads(resp.read().decode("utf-8"))

async def ensure_user(user_id: str):
    loop = asyncio.get_event_loop()
    try:
        await loop.run_in_executor(None, lambda: _make_request(
            f"{BASE_URL}/v1/users",
            {"user_id": user_id, "name": "Eternal Architect Agent", "timezone": "UTC"}
        ))
    except Exception:
        pass  # Ignore 409 or other errors

async def save_memory(
    title: str,
    content: str,
    tags: list = None,
    category: str = "General",
    source: str = "agent",
    project_id: str = None,
    description: str = None
):
    """Save a memory to Parcle with full metadata schema."""
    if tags is None:
        tags = []

    timestamp = datetime.utcnow().isoformat() + "Z"
    project_name = project_id or os.getenv("ENTER_PROJECT_ID", "eternal-architect")

    print(f"[Parcle] 💾 saveMemory → title=\"{title}\" category={category} source={source}")

    try:
        await ensure_user(DEFAULT_USER_ID)

        # Build rich tag object with all required metadata fields
        tag_object = {
            "project": project_name,
            "timestamp": timestamp,
            "category": category,
            "source": source,
            "decision": "true",
            "architect": "true",
            "title": title[:200]
        }

        if description:
            tag_object["description"] = description[:200]  # Parcle tag value limit

        for t in tags:
            if isinstance(t, str) and ':' in t:
                k, v = t.split(':', 1)
                tag_object[k.strip()] = v.strip()
            elif isinstance(t, str):
                tag_object[t] = "true"
            elif isinstance(t, dict):
                tag_object.update(t)

        payload = {
            "user_id": DEFAULT_USER_ID,
            "messages": [
                {"role": "user",      "content": f"[{category}] {title}"},
                {"role": "assistant", "content": content}
            ],
            "tag": tag_object
        }

        loop = asyncio.get_event_loop()
        data = await loop.run_in_executor(
            None,
            lambda: _make_request(f"{BASE_URL}/v1/memories/ingest_dialog", payload)
        )
        session_id = data.get("session_id")
        print(f"[Parcle] ✅ Memory saved → session_id={session_id} category={category}")
        return session_id
    except Exception as e:
        print(f"[Parcle] ❌ save_memory failed: {e}")
        return None

async def query_memory(natural_language_query: str):
    print(f"[Parcle] 🔍 queryMemory → \"{natural_language_query}\"")
    try:
        loop = asyncio.get_event_loop()

        api_key = get_api_key()
        payload = json.dumps({
            "user_id": DEFAULT_USER_ID,
            "query": natural_language_query
        }).encode("utf-8")

        req = urllib.request.Request(
            f"{BASE_URL}/v1/memories/search",
            data=payload,
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
                "Accept": "text/event-stream",
                "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/111.0"
            },
            method="POST"
        )

        def do_query():
            with urllib.request.urlopen(req, timeout=20) as resp:
                body = resp.read().decode("utf-8")
            return body

        body = await loop.run_in_executor(None, do_query)

        lines = body.splitlines()
        seen_final = False
        final_data = None
        for line in lines:
            line = line.strip()
            if line == "event: final":
                seen_final = True
                continue
            if seen_final and line.startswith("data:"):
                try:
                    final_data = json.loads(line[5:].strip())
                except Exception:
                    pass
                break

        if not final_data:
            print("[Parcle] No final event found in search response.")
            return []

        print(f"[Parcle] ✅ queryMemory result → confidence={final_data.get('confidence')} citations={len(final_data.get('citations', []))}")

        results = [{
            "query":      natural_language_query,
            "answer":     final_data.get("answer"),
            "confidence": final_data.get("confidence"),
            "citations":  final_data.get("citations", [])
        }]

        citations = final_data.get("citations", [])
        if len(citations) > 1:
            extras = [{
                "query":      natural_language_query,
                "answer":     final_data.get("answer"),
                "confidence": final_data.get("confidence"),
                "citations":  [c]
            } for c in citations[1:5]]
            results.extend(extras)

        return results[:5]
    except Exception as e:
        print(f"[Parcle] ❌ queryMemory failed: {e}")
        raise

async def list_recent_memories(limit: int = 10):
    print(f"[Parcle] 📋 listRecentMemories → limit={limit}")
    try:
        loop = asyncio.get_event_loop()
        data = await loop.run_in_executor(None, lambda: _make_request(
            f"{BASE_URL}/v1/memories/sources",
            {"user_id": DEFAULT_USER_ID, "type": "session", "order": "desc", "limit": limit, "page": 1}
        ))
        sources = data.get("sources", [])
        sources.sort(key=lambda x: x.get("updated_at") or "", reverse=True)
        return sources
    except Exception as e:
        print(f"[Parcle] ❌ list_recent_memories failed: {e}")
        return []
